### PR TITLE
Expose custom scalars configuration from CLI.

### LINKED ISSUE #2162: relay-compiler: config of customScalars for accurate flow types
I have a custom scalar on my backend called `DataUrl` which is just a string which has been validated to be a `data:` url.

In https://github.com/facebook/relay/blob/master/packages/relay-compiler/bin/RelayCompilerBin.js#L177, `customScalars` is passed as an empty object, but could be configuration-based.

It is used by https://github.com/facebook/relay/blob/master/packages/relay-compiler/core/RelayFlowTypeTransformers.js#L74. There appears to be special support for the nonstandard `Url` scalar type already.

### PR DESCRIPTION
**Pull request by alloy**
Closes #2162.

In preparation of #2518, as requested by @josephsavona we should have CLI versions of each possible argument, hereby CLI support for custom scalars.

I ended up going with the dot-notation syntax of yargs, which seems a little foreign (to me), but it comes builtin with yargs. Let me know what you think.

E.g.

```
$ relay-compiler --customScalars.URL=String --customScalars.Date=String
```

### CODE REVIEW

File: packages/relay-compiler/bin/RelayCompilerBin.js [hunk 1]
```diff
@@ -100,6 +100,14 @@ const argv = yargs
       type: 'string',
       default: null,
     },
+    customScalars: {
```
Comment (alloy) [inline packages/relay-compiler/bin/RelayCompilerBin.js]: Re case, see https://github.com/facebook/relay/issues/2518#issuecomment-496176977

File: packages/relay-compiler/bin/RelayCompilerMain.js [hunk 2]
```diff
@@ -174,86 +176,107 @@ Root files can include:
...
+    ? await codegenRunner.watchAll()
+    : await codegenRunner.compileAll();
+
+  if (result === 'ERROR') {
+    process.exit(100);
+  }
+  if (options.validate && result !== 'NO_CHANGES') {
+    process.exit(101);
+  }
+}
```
Comment (alloy) [inline packages/relay-compiler/bin/RelayCompilerMain.js]: Most of these changes are to refactor configuration of the codegen runner out into a standalone function that can be tested.

File: gulpfile.js [hunk 3]
```diff
@@ -260,6 +260,7 @@ const modules = gulp.parallel(
               '!**/__tests__/**',
               '!**/__flowtests__/**',
               '!**/__mocks__/**',
+              '!**/node_modules/**',
```
Comment (alloy) [inline gulpfile.js]: I lost a bunch of time faffing around trying to find the best way to do continuous development on the package while integrated into one of our apps, which led me to at some point running `yarn install` in the `packages/relay-compiler` dir. After that building failed in confusing ways, so this addition is really just to guard other devs from the same silly mistake.

**Comment by alloy** [comment]
Note that I had to do some synthetic tests, as none of our schema’s custom scalars really maps to a single builtin GraphQL scalar type, but doing so did update the [typescript] artifacts in our app as expected.

For instance, we have a `FormattedNumber` scalar, which holds either a string or a number, but afaict the existing custom scalar support does not allow something like `{ "FormattedNumber": "String | Int" }`. Is that correct?

**Comment by TrySound** [comment]
Awesome! I want it so much. Currently to make date/time scalars type safe I have to copy/paste RelayCompilerBin and RelayCompilerMain into my project.

File: packages/relay-compiler/bin/RelayCompilerBin.js [hunk 4]
```diff
@@ -100,6 +100,14 @@ const argv = yargs
...
+      describe:
+        'Mappings from custom scalars in your schema to built-in GraphQL ' +
+        'types, for type emission purposes. (Uses yargs dot-notation, e.g. ' +
+        '--customScalars.URL=String)',
+      type: 'object',
```
Comment (jstejada) [inline packages/relay-compiler/bin/RelayCompilerBin.js]: @alloy I'm getting flow errors for this type. Is it actually supported? I can't find it in the docs: http://yargs.js.org/docs/#api-optionkey-opt

Comment (jstejada) [reply packages/relay-compiler/bin/RelayCompilerBin.js]: going to merge this in anyway and we can fix in a follow up

Comment (alloy) [reply packages/relay-compiler/bin/RelayCompilerBin.js]: It doesn’t exist [yet](https://github.com/yargs/yargs/issues/890), in the sense that it checks the input type, but setting it here does make it show up in the help banner and so I think it’s helpful to include already.

I did not notice any flow errors at the time, but also the relay-config PR applies custom typings to these options, so perhaps that overrides any current failures here.

Comment (jstejada) [reply packages/relay-compiler/bin/RelayCompilerBin.js]: got it! thanks! I'll just leave it with a $FlowFixMe for now then

**Pull request closed without merging**